checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -368,52 +455,82 @@ private void analyzeOwnershipTransferAtAssignment(
+        addMustCallAliasToFormalParameter(paramIndex);
+        mcca.removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);
+      } else {
+        addOwningToParam(paramIndex);
+        mcca.removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);
+      }
+
     } else if (lhsElement.getKind() == ElementKind.RESOURCE_VARIABLE && mcca.isMustCallClose(rhs)) {
-      addOwningToParamsIfDisposedAtAssignment(obligations, rhsObligation, rhs);
+      int paramIndex = getIndexOfParam(rhsObligation);
+      if (paramIndex > 0) {
+        addOwningToParam(paramIndex);
+      }
     } else if (lhs instanceof LocalVariableNode) {
       LocalVariableNode lhsVar = (LocalVariableNode) lhs;
       mcca.updateObligationsForPseudoAssignment(obligations, assignmentNode, lhsVar, rhs);
     }
   }
 
   /**
-   * If a must-call obligation of some alias of method parameter p is satisfied during the
-   * assignment, add an @Owning annotation to p, and remove the rhs node from the obligations set,
-   * since it no longer needs to be tracked.
+   * Return the (1-based) index of the method parameter that is an alias of the given {@code
+   * obligation}, if one exists; otherwise, return -1.
    *
-   * @param obligations the set of obligations to update
-   * @param rhsObligation the obligation associated with the right-hand side of the assignment
-   * @param rhs the right-hand side of the assignment
+   * @param obligation the obligation
+   * @return the index of the current method parameter that is must-aliased to the given obligation,
+   *     if one exists; otherwise, return -1.
    */
-  private void addOwningToParamsIfDisposedAtAssignment(
-      Set<Obligation> obligations, Obligation rhsObligation, Node rhs) {
-    Set<ResourceAlias> rhsAliases = rhsObligation.resourceAliases;
-    if (rhsAliases.isEmpty()) {
-      return;
-    }
+  private int getIndexOfParam(Obligation obligation) {
+    Set<ResourceAlias> resourceAliases = obligation.resourceAliases;
     List<VariableElement> paramElts =
         CollectionsPlume.mapList(TreeUtils::elementFromDeclaration, methodTree.getParameters());
-    for (ResourceAlias rhsAlias : rhsAliases) {
-      Element rhsElt = rhsAlias.element;
-      int i = paramElts.indexOf(rhsElt);
-      if (i != -1) {
-        addOwningToParam(i + 1);
-        mcca.removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);
-        break;
+    for (ResourceAlias resourceAlias : resourceAliases) {
+      int paramIndex = paramElts.indexOf(resourceAlias.element);
+      if (paramIndex != -1) {
+        return paramIndex + 1;
       }
     }
+
+    return -1;
+  }
+
+  /** Adds a {@link NotOwning} annotation to the return type of the current method. */
+  private void addNotOwningToReturnType() {
+    WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+    wpi.addMethodDeclarationAnnotation(methodElt, NOTOWNING);
```

Comment:
Writing a *method declaration* here is inconsistent with the documentation that says the annotation is added to the *return type*.